### PR TITLE
[HxInputBase] Allow for CustomValidation Context and Propety and Add LabelType into application-wide Defaults

### PR DESCRIPTION
**Pull request by mmonteagudo**
Linked Issue
#872

### CODE REVIEW

**Comment by mmonteagudo** [comment]
https://github.com/havit/Havit.Blazor/issues/872

File: Havit.Blazor.Components.Web.Bootstrap/Forms/InputSettings.cs [hunk 1]
```diff
@@ -9,4 +9,8 @@ public record InputSettings
 	/// Specifies how the validation message should be displayed.
 	/// </summary>
 	public ValidationMessageMode? ValidationMessageMode { get; set; }
+	/// <summary>
+	/// Specifies how the label should be displayed.
+	/// </summary>
+	public LabelType? LabelType { get; set; }
```
Comment (hakenr) [inline Havit.Blazor.Components.Web.Bootstrap/Forms/InputSettings.cs:15]: The current issue is that not all input components have `LabelType` (implement `IInputWithLabelType`), which is why we left this unimplemented and didn't allow setting `LabelType` from a single place.

However, I understand there's a valid need to have a central point to set `LabelType`, even if some input components don't support it. I'll do some rework here...

File: Havit.Blazor.Documentation/Pages/Components/FormInputs/FormInputs_Demo_CustomValidation.razor [hunk 2]
```diff
@@ -0,0 +1,65 @@
+﻿<EditForm Model="@_testField" OnValidSubmit="@(() => model.Text = "Validation passed... :-)")">
+	<DataAnnotationsValidator />
+
+    <HxInputText Label="HxInputText" @bind-Value="@_testField.TestText" ValidationEditContext="new(_testField.TestModel)" ValidationFor="() => _testField.TestModel.Text"
+        Placeholder="Enter text here." Hint="Hint: Enter text, the value is shared with all text inputs." />
```
Comment (hakenr) [inline Havit.Blazor.Documentation/Pages/Components/FormInputs/FormInputs_Demo_CustomValidation.razor:5]: This seems a bit unclear:

* The current `<EditForm />` uses `_testField` as the `Model`, while the inputs bind to `@model` properties. Why not use `<EditForm Model="@model">` instead?
* The new `ValidationXy` will cause the form to display a validation message that the user cannot resolve. Is there no way to set `_testField.TestModel.Text`? Additionally, the `HxInputText` will still trigger validation on the bound property, even if the `HxValidationMessage` is redirected elsewhere.
* Have you considered using nested `<EditForm />` components for the different models used by `HxInputText` in this context?
* Have you thought about using `<HxValidationMessage />` directly, combined with `<HxInputText ValidationMessageMode="ValidationMessageMode.None" />` for inputs where you don't want to expose their validation messages?

@mmonteagudo I might not fully understand the intent here. Could you provide a real-world scenario to clarify?

Comment (mmonteagudo) [reply Havit.Blazor.Documentation/Pages/Components/FormInputs/FormInputs_Demo_CustomValidation.razor:5]: I want able to provide it for the examples in the code, but in my case I'm using it for Validating using Fluent Validation when validating values that are not parameters, for example in my case:

One if my properties in my project is ApplicationData which is a Dictionary, and I need to validate when one for the items in the dictionary has an specific value as a condition to validate all others, as default validation only works with properties, that's not possible, so in this in case I declared in Fluent Validation a "custom property to validate" and I'm matching that property (in this case equals to the dictionary key)

So, basically I'm sending to Fluent a custom property name to validate.

Hope the explanation helps, here are some code examples where I'm using the implementation

image

image

Comment (hakenr) [reply Havit.Blazor.Documentation/Pages/Components/FormInputs/FormInputs_Demo_CustomValidation.razor]: Custom validation feature removed from the PR.

**Comment by hakenr** [comment]
Input components (`HxInputBase` descendants) that do not support `LabelType`:
* `HxCheckbox`, `HxSwitch`
* `HxCheckboxList`, `HxRadioButtonList`
* `HxInputDateRange`
* `HxInputRange`
* `HxInputSelect` - can we add support here?

...Adding `LabelType` to `InputSettings` will introduce this property to the settings/defaults of all inputs, which could lead to some confusion. I'm considering an alternative approach by adding a brand new ambient defaults such as `HxSetup.Defaults.LabelType` and `HxSetup.Defaults.InputSize`. I'll discuss this with my colleagues.

**Review by hakenr — APPROVED** [review]

**Comment by hakenr** [comment]
@mmonteagudo Thanks a lot for your contribution! It made me further develop the idea of "global defaults for all inputs," and I’ve added `HxSetup.Defaults` for these global settings, which currently allow setting `LabelType` and `InputSize`.

Apologies for not accepting the custom validation parameters idea. I’m concerned that this feature could introduce more issues than the value it adds. We might revisit this later, but for now, I’ve decided against it, as there are possible workarounds that can help you achieve the same goal without the proposed changes.

**Comment by mmonteagudo** [comment]
> @mmonteagudo Thanks a lot for your contribution! It made me further develop the idea of "global defaults for all inputs," and I’ve added `HxSetup.Defaults` for these global settings, which currently allow setting `LabelType` and `InputSize`.
> 
> Apologies for not accepting the custom validation parameters idea. I’m concerned that this feature could introduce more issues than the value it adds. We might revisit this later, but for now, I’ve decided against it, as there are possible workarounds that can help you achieve the same goal without the proposed changes.

No worries, I understand.

That's an important feature for our implementation, so we are using it from our fork of Havit, so we just keep it up to date with latest updates and deploy a new package in our private package manager.

I do have a couple of customizations that works for our specific implementation, so I understand. :), Thanks for accepting the other change, that one was pretty important too.
